eoe/man/man1/tabs.1
'\"macro stdmacro
.if n .pH g1.tabs @(#)tabs	30.15 of 3/22/86
.nr X
.if \nX=0 .ds x} TABS 1 "User Environment Utilities" "\&"
.if \nX=1 .ds x} TABS 1 "User Environment Utilities"
.if \nX=2 .ds x} TABS 1 "" "\&"
.if \nX=3 .ds x} TABS "" "" "\&"
.TH \*(x}
.SH NAME
tabs \- set tabs on a terminal
.SH SYNOPSIS
.BR tabs
[tabspec]
.RB [ \-T "type]"
.RB [ +m n]
.SH DESCRIPTION
.I tabs\^
sets the tab stops on the user's terminal according to the tab specification
.IR tabspec ,
after clearing any previous settings.
The user's terminal must have remotely-settable hardware tabs.
.PP
.TP 8
.I tabspec
Four types of tab specification are accepted for
.IR tabspec .
They are described below:
canned
(\f3\-\f2code\f1), repetitive (\f3\-\f2n\f1),
arbitrary (\f2n1,n2,...\f1),
and file (\f3\-\-\f2file\f1).
If no
.I tabspec\^
is given, the default value is
.BR \-8 ,
i.e.,
\s-1UNIX\s+1 system ``standard'' tabs.
The lowest column number is 1.
Note that for
.IR tabs ,
column 1 always refers to the leftmost column on a terminal,
even one whose column markers
begin at 0,
e.g., the \s-1DASI\s+1 300, \s-1DASI\s+1 300s, and \s-1DASI\s+1 450.
.PP
.TP 8
.BI \- code\^
Use one of the codes listed below to select a
\f2canned\f1 set of tabs.
The legal codes and their meanings are as follows:
.RS 8
.PD 0
.TP 8
.B \-a
1,10,16,36,72
.br
Assembler, \s-1IBM\s+1 S/370, first format
.TP
.B \-a2
1,10,16,40,72
.br
Assembler, \s-1IBM\s+1 S/370, second format
.TP
.B \-c
1,8,12,16,20,55
.br
\s-1COBOL\s+1, normal format
.TP
.B \-c2
1,6,10,14,49
.br
\s-1COBOL\s+1 compact format (columns 1-6 omitted).
Using this code, the first typed character corresponds to card column 7,
one space gets you to column 8, and a tab reaches column 12.
Files using this tab setup
should include a format specification
as follows (see \f2fspec\f1(4)):
.in +.5i
.B "<:t\-c2 \|m6 \|s66 \|d:>"
.in -.5i
.TP
.B \-c3
1,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,67
.br
\s-1COBOL\s+1 compact format (columns 1-6 omitted), with more tabs than
.B \-c2.
This is the recommended format for \s-1COBOL\s+1.
The appropriate format specification is (see \f2fspec\f1(4)):
.in +.5i
.B "<:t\-c3 \|m6 \|s66 \|d:>"
.in -.5i
.TP
.B \-f
1,7,11,15,19,23
.br
\s-1FORTRAN\s+1
.TP
.B \-p
1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61
.br
\s-1PL/I\s+1
.TP
.B \-s
1,10,55
.br
\s-1SNOBOL\s+1
.TP
.B \-u
1,12,20,44
.br
\s-1UNIVAC\s+1 1100 Assembler
.RE
.PD
.br
.ne 10
.TP 8
.BI \- n\^
A \f2repetitive\f1 specification requests tabs at columns
.RI 1+ n ,
.RI 1+2\(** n ,
etc.
Of particular importance is the value
.BR 8 :
this represents the
\s-1UNIX\s+1 system ``standard'' tab setting, and is the most likely tab setting to be found
at a terminal.
Another special case is the value
.BR 0 ,
implying no tabs at all.
.TP
.IR n1 , n2 ,...
The \f2arbitrary\f1 format permits the user to type any
chosen set of numbers, separated by commas, in ascending order.
Up to 40 numbers are allowed.
If any number (except the first one) is preceded by a plus sign, it is taken
as an increment to be added to the previous value.
Thus, the formats
\f31\f1,\f310\f1,\f320\f1,\f330\f1,
and
\f31\f1,\f310\f1,\f3+10\f1,\f3+10\f1
are considered identical.
.TP
.BI \-\- file\^
If the name of a \f2file\f1 is given,
.I tabs\^
reads the first line of the file, searching for a format specification (see
\f2fspec\f1(4)).
If it finds one there, it sets the tab stops according to it, otherwise it
sets them as
.BR \-8 .
This type of specification may be used to make sure that a tabbed file is printed
with correct tab settings, and would be used with the
.IR pr (1)
command:
.in +.5i
\f3tabs \-\-\f1 file; \f3pr\f1 file
.in -.5i
.RE
.PP
Any of the following also may be used;
if a given flag occurs more than once,
the last value given takes effect:
.PP
.TP 8
.BI \-T type\^
.I tabs\^
usually needs to know the type of terminal in order to set tabs
and always needs to know the type to set margins.
.I type\^
is a name listed in
.IR term (5).
If no
.B \-T
flag is supplied,
.I tabs\^
uses the value of the environment variable
.BR \s-1TERM\s0 .
If
.B \s-1TERM\s0
is not defined
in the
.I environment\^
(see
.IR environ (5)),
.I tabs\^
tries a sequence that will work for many terminals.
.TP
.BI +m n\^
The margin argument may be used for some terminals.
It causes
all tabs to be moved over
.I n\^
columns by making column
.I n+1\^
the left margin.
If
.B +m
is given without a value of
.IR n ,
the value assumed is
.BR 10 .
For a TermiNet, the first value in the tab list should
be
.BR 1 ,
or the margin will move even further to the right.
The normal (leftmost) margin on most terminals is
obtained by
.BR +m0 .
The margin for most terminals is reset only when the
.B +m
flag is given explicitly.
.SH EXAMPLES
.TP 13
.B "tabs \-a"
example using
.I \-code
(\f2canned\f1 specification) to set
tabs to the settings required by the IBM
assembler:  columns 1, 10, 16, 36, 72.
.TP
.B "tabs \-8"
example of using
.I \-n
(\f2repetitive\f1 specification), where
.I n
is
.BR 8 ,
causes tabs to be set every eighth position:
.br
1+(1*8), 1+(2*8), ... which evaluate to columns
9, 17, ...
.TP
.B "tabs 1,8,36"
example of using
.IR n1 , n2 ,...
(\f2arbitrary\f1 specification)
to set tabs at columns 1, 8, and 36.
.br
.ne 10
.TP
.B "tabs \-\-$HOME/fspec.list/att4425"
example of using
.I \-\-file
(\f2file\f1 specification) to indicate
that tabs should be set according to the first
line of
.I $HOME/fspec.list/att4425
(see \f2fspec\f1(4)).
.br
.ne 10
.SH DIAGNOSTICS
.PD 0
.TP "\w@\f2unknown\ tab\ code\fP\ \ \ \ @u"
.I "illegal tabs\^"
when arbitrary tabs are ordered incorrectly
.br
.TP
.I "illegal increment\^"
when a zero or missing increment is found in
an arbitrary specification
.br
.TP
.I "unknown tab code\^"
when a \f2canned\f1 code cannot be found
.br
.TP
.I "can't open\^"
if \f3\-\-\fP\f2file\fP option used, and file can't be opened
.br
.TP
.I "file indirection\^"
if \f3\-\-\fP\f2file\fP option used and the specification
in that file points to yet another file.
Indirection of this form is not permitted
.br
.PD
.SH NOTES
Hardware tabs must be enabled on the terminal device by entering the UNIX
command \f2`stty tabs'\f1;
otherwise the
.I tabs
command will appear to have no effect.
Tab and margin setting is performed via the standard output.
.PP
There is no consistency among different terminals regarding ways of
clearing tabs and setting the left margin.
.P
.I tabs\^
clears only 20 tabs (on terminals requiring a long sequence),
but is willing to set 64.
.SH WARNING
The
.I tabspec
used with the
.I tabs
command is different from the one used with the
.IR newform (1)
command.
For example,
.B "tabs \-8"
sets every eighth position;
whereas
.B "newform \-i\-8"
indicates that tabs are set every eighth position.
.SH SEE ALSO
stty(1), newform(1), pr(1), tput(1),
fspec(4),
terminfo(4),
environ(5), term(5)
.\"	@(#)tabs.1	6.2 of 9/2/83
.Ee
'\".so /pubs/tools/origin.att


eoe/cmd/tabs/tabs.c
/*	Copyright (c) 1984 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)tabs:tabs.c	1.8"

/*
 *	tabs [tabspec] [+mn] [-Ttype]
 *	set tabs (and margin, if +mn), for terminal type
 */


#include <stdio.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <curses.h>
#include <term.h>
#define EQ(a,b)	(strcmp(a, b) == 0)
/*	max # columns used (needed for GSI) */
#define NCOLS	256
#define NTABS	65	/* max # tabs +1 (to be set) */
#define	NTABSCL	21	/* max # tabs + 1 that will be cleared */
#define ESC	033
#define CLEAR	'2'
#define SET	'1'
#define TAB	'\t'
#define CR	'\r'
#define NMG	0	/* no margin setting */
#define GMG	1	/* DTC300s margin */
#define TMG	2	/* TERMINET margin */
#define DMG	3	/* DASI450 margin */
#define FMG	4	/* TTY 43 margin */
#define TRMG	5	/* Trendata 4000a */

#define TCLRLN	0	/* long, repetitive, general tab clear */

int	err;
int 	tmarg, getmarg();
int 	ttp;
char	settab[32], clear_tabs[32];

int	maxtab;		/* max tab for repetitive spec */
int	margin;	
int	margflg;	/* >0 ==> +m option used, 0 ==> not */
char	*terminal = "";
char	*tabspec = "-8";	/* default tab specification */

struct termio ttyold;	/* tty table */
int	ttyisave;	/* save for input modes */
int	ttyosave;	/* save for output modes */
int	istty;		/* 1 ==> is actual tty */

struct	stat	statbuf;
char	*devtty;

void	endup();
char	*getenv();

main(argc, argv)
char **argv;
{
	int tabvect[NTABS];	/* build tab list here */
	int need_term = 0;
	char *ttyname();
	char *scan;	/* scan pointer to next char */

	signal(SIGINT, endup);
	if (ioctl(1, TCGETA, &ttyold) == 0) {
		ttyisave = ttyold.c_iflag;
		ttyosave = ttyold.c_oflag;
		fstat(1, &statbuf);
		devtty = ttyname(1);
		chmod(devtty, 0000);	/* nobody, not even us */
		istty++;
	}
	tabvect[0] = 0;	/* mark as not yet filled in */
	while (--argc > 0) {
		scan = *++argv;
		if (*scan == '+')
			switch (*++scan) {
			case 'm':
				margflg++;
				if (*++scan)
					margin = getnum(&scan);
				else
					margin = 10;
				break;
			}
		else if (*scan == '-' && *(scan+1) == 'T') {
			terminal = scan+2;
			need_term++;
		} else {
			tabspec = scan;		/* save tab specification */
		}
	}
	if (*terminal == '\0') {
	    if (need_term) {
		fprintf(stderr, "tabs: Need to specify a terminal type for the -T option\n");
		exit(1);
	    } else 
		terminal = getenv("TERM");
	}

	if ((terminal == NULL) || (*terminal == '\0')) {
		fprintf(stderr, "tabs: No value for $TERM and no -T specified\n");
		exit(1);
	}

	setupterm(terminal,1,&err);
	if (err <= 0) {
		fprintf(stderr,
			"tabs: Could not find `%s' in terminfo data base\n",
			terminal);
		exit(1);
	}
	if (set_tab == NULL) {
		fprintf(stderr,
		 "tabs: terminal type `%s' does not have the ability to set tabs\n",
			terminal);
		exit(1);
	}
	maxtab = columns;
	strcpy(settab,set_tab);
	strcpy(clear_tabs,clear_all_tabs);

	scantab(tabspec,tabvect,0);
	if (!tabvect[0])
		repetab("8",tabvect);
		settabs(tabvect);
	endup();
	exit(0);
}

/*	scantab: scan 1 tabspec & return tab list for it */

scantab(scan,tabvect,level)
char *scan;
int tabvect[NTABS], level;
{
	register char c;
	if (*scan == '-')
		if ((c = *++scan) == '-')
			filetab(++scan,tabvect,level);
		else if (c >= '0' && c <= '9')
			repetab(scan,tabvect);
		else if (stdtab(scan,tabvect))
			error("unknown tab code");
		else;
	else
		arbitab(scan,tabvect);
}

/*	repetab: scan and set repetitve tabs, 1+n, 1+2*n, etc */

repetab(scan,tabvect)
char *scan;
int tabvect[NTABS];
{
	register incr, i, tabn;
	int limit;
	incr = getnum(&scan);
	tabn = 1;
	limit = (maxtab-1)/(incr?incr:1)-1; /* # last actual tab */
	if (limit>NTABS-2)
		limit = NTABS-2;
	for (i = 0; i<=limit; i++)
		tabvect[i] = tabn += incr;
	tabvect[i] = 0;
}

/*	arbitab: handle list of arbitrary tabs */

arbitab(scan,tabvect)
char *scan;
int tabvect[NTABS];
{
	register i, t, last;
	last = 0;
	for (i = 0; i<NTABS-1;) {
		if (*scan == '+') {
			scan++;		/* +n ==> increment, not absolute */
			if (t = getnum(&scan))
				tabvect[i++] = last += t;
			else error("illegal increment");
		}
		else {
			if ((t = getnum(&scan)) > last)
				tabvect[i++] = last = t;
			else error("illegal tabs");
		}
		if (*scan++ != ',') break;
	}
	if (last > NCOLS)
		error("illegal tabs");
	tabvect[i] = 0;
}

/*	filetab: copy tabspec from existing file */
#define CARDSIZ	132
filetab(scan,tabvect,level)
char *scan;
int tabvect[NTABS];
{
	register length, i;
	register char c;
	int fildes;
	char card[CARDSIZ];	/* buffer area for 1st card in file */
	char state, found;
	char *temp;
	if (level)
		error("file indirection");
	if ((fildes = open(scan,0)) < 0)
		error("can't open");
	length = read(fildes,card,CARDSIZ);
	close(fildes);
	found = state = 0;
	scan = 0;
	for (i = 0; i<length && (c = card[i]) != '\n'; i++) {
		switch (state) {
		case 0:
			state = (c == '<'); break;
		case 1:
			state = (c == ':')?2:0; break;
		case 2:
			if (c == 't')
				state = 3;
			else if (c == ':')
				state = 6;
			else if (c != ' ')
				state = 5;
			break;
		case 3:
			if (c == ' ')
				state = 2;
			else {
				scan = &card[i];
				state = 4;
			}
			break;
		case 4:
			if (c == ' ') {
				card[i] = '\0';
				state = 5;
			}
			else if (c == ':') {
				card[i] = '\0';
				state = 6;
			}
			break;
		case 5:
			if (c == ' ')
				state = 2;
			else if (c == ':')
				state = 6;
			break;
		case 6:
			if (c == '>') {
				found = 1;
				goto done;
			}
			else state = 5;
			break;
		}
	}
done:
	if (found && scan != 0) {
		scantab(scan,tabvect,1);
		temp = scan;
		while (*++temp);
		*temp = '\n';
	}
	else scantab("-8",tabvect,1);
}

int
getmarg (term)
char *term;
{
	if (strncmp(term,"1620",4) == 0 || strncmp(term,"1700",4) == 0 || strncmp(term,"450",3) == 0)
		return(DMG);
	else if (strncmp(term,"300s",4) == 0)
		return(GMG);
	else if (strncmp(term,"4000a",5) == 0)
		return(TRMG);
	else if (strcmp(term,"43") == 0)
		return(FMG);
	else if (strcmp(term,"tn300") == 0 || strcmp(term,"tn1200") == 0)
		return(TMG);
	else return(NMG);
}

char	*cleartabs();
/*	settabs: set actual tabs at terminal */
/*	note: this code caters to necessities of handling GSI and
	other terminals in a consistent way. */

settabs(tabvect)
int tabvect[NTABS];
{
	char setbuf[512];	/* 2+3*NTABS+2+NCOLS+NTABS (+ some extra) */
	register char *p;		/* ptr for assembly in setbuf */
	register *curtab;		/* ptr to tabvect item */
	int i, previous, nblanks;
	if (istty) {
		ttyold.c_iflag &= ~ICRNL ;
		ttyold.c_oflag &= ~(ONLCR|OCRNL|ONOCR|ONLRET);
		ioctl(1, TCSETAW, &ttyold);	/* turn off cr-lf map */
	}
	p = setbuf;
	*p++ = CR;
	p = cleartabs(p, clear_tabs);

	if (margflg) {
		tmarg = getmarg(terminal);
		switch(tmarg) {
		case GMG:	/* GSI300S */
		/* NOTE: the 300S appears somewhat odd, in that there is
		a column 0, but there is no way to do a direct tab to it.
		The sequence ESC 'T' '\0' jumps to column 27 and prints
		a '0', without changing the margin. */
			*p++ = ESC;
			*p++ = 'T';	/* setup for direct tab */
			if (margin &= 0177)	/* normal case */
				*p++ = margin;
			else {			/* +m0 case */
				*p++ = 1;	/* column 1 */
				*p++ = '\b';	/* column 0 */
			}
			*p++ = margin;	/* direct horizontal tab */
			*p++ = ESC;
			*p++ = '0';	/* actual margin set */
			break;
		case TMG:	/* TERMINET 300 & 1200 */
			while (margin--)
				*p++ = ' ';
			break;
		case DMG:	/* DASI450/DIABLO 1620 */
			*p++ = ESC;	/* direct tab ignores margin */
			*p++ = '\t';
			if (margin == 3){
				*p++ = (margin & 0177);
				*p++ = ' ';
			}
			else
				*p++ = (margin & 0177) + 1;
			*p++ = ESC;
			*p++ = '9';
			break;
		case FMG:	/* TTY 43 */
			p--;
			*p++ = ESC;
			*p++ = 'x';
			*p++ = CR;
			while (margin--)
				*p++ = ' ';
			*p++ = ESC;
			*p++ = 'l';
			*p++ = CR;
			write(1, setbuf, p - setbuf);
			return;
		case TRMG:
			p--;
			*p++ = ESC;
			*p++ = 'N';
			while (margin--)
				*p++ = ' ';
			*p++ = ESC;
			*p++ = 'F';
			break;
		}
	}

/*
 *	actual setting: at least terminals do this consistently!
 */
	previous = 1; curtab = tabvect;
	while ((nblanks = *curtab-previous) >= 0 &&
		previous + nblanks <= maxtab) {
		for (i = 1; i <= nblanks; i++) *p++ = ' ';
		previous = *curtab++;
		strcpy(p,settab);
		p += strlen(settab);
	}
	*p++ = CR;
	if (EQ(terminal,"4424"))
		*p++ = '\n';	/* TTY40/2 needs LF, not just CR */
	write(1, setbuf, p - setbuf);
}

/*	cleartabs(pointer to buffer, pointer to clear sequence) */
char *cleartabs(p, qq)
register char *p;
char *qq;
{
	register i;
	register char *q;
	q = qq;
	if (clear_tabs == 0) {	      /* if repetitive sequence */
		*p++ = CR;
		for(i = 0; i < NTABSCL - 1; i++) {
			*p++ = TAB;
			*p++ = ESC;
			*p++ = CLEAR;
		}
		*p++ = CR;
	}
	else {
		while(*p++ = *q++);	/* copy table sequence */
		p--;			/* adjust for null */
		if (EQ(terminal,"4424")) {	/* TTY40 extra delays needed */
			*p++ = '\0';
			*p++ = '\0';
			*p++ = '\0';
			*p++ = '\0';
		}
	}
	return(p);
}
/*	getnum: scan and convert number, return zero if none found */
/*	set scan ptr to addr of ending delimeter */
getnum(scan1)
char **scan1;
{
	register n;
	register char c, *scan;
	n = 0;
	scan = *scan1;
	while ((c = *scan++) >= '0' && c <= '9') n = n * 10 + c -'0';
	*scan1 = --scan;
	return(n);
}

/*	error: terminate processing with message to terminal */
error(arg)
char *arg;
{
	register char *temp;
	temp = arg;
	while (*++temp);	/* get length */
	*temp = '\n';
	endup();
	write(2, arg, temp+1-arg);
	exit(1);
}

/*	endup: make sure tty mode reset & exit */
void
endup()
{
	if (istty) {
		ttyold.c_iflag = ttyisave;
		ttyold.c_oflag = ttyosave;
		ioctl(1, TCSETAW, &ttyold);	/* reset cr-lf to previous */
		chmod(devtty, statbuf.st_mode & 0777);
	}
	if (err > 0)
		resetterm();
}

/*	stdtabs: standard tabs table
	format: option code letter(s), null, tabs, null */
char stdtabs[] = {
'a',	0,1,10,16,36,72,0,			/* IBM 370 Assembler */
'a','2',0,1,10,16,40,72,0,			/* IBM Assembler alternative*/
'c',	0,1,8,12,16,20,55,0,			/* COBOL, normal */
'c','2',0,1,6,10,14,49,0,			/* COBOL, crunched*/
'c','3',0,1,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,67,0,
						/* crunched COBOL, many tabs */
'f',	0,1,7,11,15,19,23,0,			/* FORTRAN */
'p',	0,1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61,0, /* PL/I */
's',	0,1,10,55,0,				/* SNOBOL */
'u',	0,1,12,20,44,0,				/* UNIVAC ASM */
0};

/*	stdtab: return tab list for any "canned" tab option.
	entry: option points to null-terminated option string
		tabvect points to vector to be filled in
	exit: return(0) if legal, tabvect filled, ending with zero
		return(-1) if unknown option
*/
stdtab(option,tabvect)
char option[];
int tabvect[];
{
	register char *sp;
	tabvect[0] = 0;
	sp = stdtabs;
	while (*sp) {
		if (EQ(option,sp)) {
			while (*sp++);		/* skip to 1st tab value */
			while (*tabvect++ = *sp++);	/* copy, make int */
			return(0);
		}
		while(*sp++);	/* skip to 1st tab value */
		while(*sp++);		/* skip over tab list */
	}
	return(-1);
}
